### DISCUSSION TITLE
Unable to get module working...

### DISCUSSION THREAD

**Discussion by rajasekarshanmugam** [sebastienros/jint#1379] [General]
Hi, Trying to use the latest version of JINT and module support looked interesting for some workarounds we had done. However when trying to define a module and call into it, unable to get a trivial sample working. Checked module tests, but none seems to have a hint on how to execute it. Am I missing something, please help?

```csharp

var engineBase = new Engine();
string math = @"
	function add(a, b) {
		return a + b;
	}
	export { add };";

engineBase.AddModule("math", math);
engineBase.ImportModule("math");
var result = engineBase.Execute(
"""
	import {add} from 'math';
	add(1,2);
"""
	);
Console.WriteLine(result);

```

- importing the module using import causes an error that import cannot be used
- if invoke is changed to add, then function is not defined
- if invoke is changed to use fully qualified math.add, then math is not defined
- if the definition is moved a folder js/math.js, engine changed to use modules loader, it still does not find the module.

```csharp

var baseJSPath = Path.Combine(Path.GetDirectoryName(typeof(Program).Assembly.Location)!, "js");
var e = new Engine(options => options.EnableModules(baseJSPath, false));

e.ImportModule("math");
var result = e.Execute(
"""
	add(1,2);
"""
	);
Console.WriteLine(result);

```

Am I missing something? Could not find any valid sample to work with, so had to try options, still could not have  break through, any help would be great.

**Comment by lahma** [comment]
I think the basic problem here that you are using non-module code to call `import`. If you want to use `import` you need to be evaluating a module. `engine.Execute()` and `engine.Evaluate()` are running in normal script context.

**Reply by rajasekarshanmugam** [reply]
Could not find a good sample, have shared a couple below that worked. Hope it helps someone looking for the same.

Option1 - load module inline
```csharp
using Jint;

var engine = new Engine();

string mathCode = """
function add(a, b)
{
    return a + b;
}
export { add };
""";
engine.AddModule("math", mathCode);
var math = engine.ImportModule("math");

var fun = math.Get("add").AsFunctionInstance();

// invoke once
var result = fun.Invoke(engine, 1, 2);
Console.WriteLine(result);

// parallel test - throws exception if not protected using any primitive
Parallel.For(0, 100, (index) =>
{
	var result2 = 0d;
	lock (fun)
	{
		result2 = fun.Invoke(engine, 1, 2).AsNumber();
	}

	Console.WriteLine(result2);
});
```

Option2 - load using external file system. Assuming "js" folder contains the math.js file

```javascript
// math.js
function add(x, y) {
	return x + y;
}

function subtract(x, y) {
	return x - y;
}

export {
	add, subtract
};
```

```csharp
using Jint;
using Jint.Runtime.Modules;

var baseJSPath = Path.Combine(Path.GetDirectoryName(typeof(Program).Assembly.Location)!, "js");
var engine = new Engine(options => options.EnableModules(baseJSPath, false));

var math = engine.ImportModule("./math.js");
var fun = math.Get("add").AsFunctionInstance();

// invoke once
var result = fun.Invoke(engine, 1, 2);
Console.WriteLine(result);

// parallel test - throws exception if not protected using any primitive
Parallel.For(0, 100, (index) =>
{
	var result2 = 0d;
	lock (fun)
	{
		result2 = fun.Invoke(engine, 1, 2).AsNumber();
	}

	Console.WriteLine(result2);
});


```

**Reply by rajasekarshanmugam** [reply]
@lahma  - a quick question. function invoke is marked obsolete, and is redirecting to engine.Invoke. Could never get that invoke working, for the above sample, what would then engine.Invoke signature look like? any pointers will help since the function.Invoke is marked as obsolete suggesting it will not be supported. Please help when possible.

**Reply by lahma** [reply]
I personally would try to avoid getting JS instances out of engine and rely on engine evaluation (this might be just me). So basically maybe trying some dynamic import ( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import ) which doesn't require module scope.

And then evaluating inside the engine like `var result = engine.Evaluate($"const func = import... ; func({myVar1}, {myVar2}})).AsNumber();` 

Just pseudo-code and ideas, haven't tested.

**Reply by rajasekarshanmugam** [reply]
Tried a few combinations, how ever "import" can never be used outside of a module and there is an error in "Evaluate" stating the same. The recommendation is good, but wondering how to achieve that - if there are any samples/tests, could certainly try and share feedback. At the moment using the obsolete method, wondering if the CI/CD pipeline is going to scream about it 😭
